aten/src/ATen/native/cuda/ConvolutionMM2d.cu

```diff
@@ -1,12 +1,525 @@
+    return *weight;
+  }
+  const auto w_sizes = weight->sizes();
+  int64_t s1 = w_sizes[0];
+  int64_t s2 = c10::multiply_integers(w_sizes.slice(1));
+  return weight->view({s1, s2});
+}
+
+void slow_conv2d_forward(
+           const Tensor &input_,
+           const Tensor &output_,
+           const Tensor &weight_,
+           const Tensor &bias,
+           const Tensor &columns,
+           const Tensor &ones_,
+           int64_t kH, int64_t kW,
+           int64_t dH, int64_t dW,
+           int64_t padH, int64_t padW) {
+  auto weight = new_view_weight_MM2d(weight_);
+  slow_conv2d_shape_check(
+      input_, {}, weight, bias, kH, kW, dH, dW, padH, padW, /*weight_nullable*/false);
+
+  TORCH_CHECK(!bias.defined() || bias.is_contiguous(),
+              "bias tensor has to be contiguous");
+
+  int ndim = input_.dim();
+  int dimf = 0;
+  int dimh = 1;
+  int dimw = 2;
+
+  if (ndim == 4) {
+    dimf++;
+    dimh++;
+    dimw++;
+  }
+
+  auto in_sizes = input_.sizes();
+  int64_t nInputPlane  = in_sizes[dimf];
+  int64_t inputHeight  = in_sizes[dimh];
+  int64_t inputWidth   = in_sizes[dimw];
+  int64_t nOutputPlane = weight.sizes()[0];
+  int64_t outputHeight = (inputHeight + 2*padH - kH) / dH + 1;
+  int64_t outputWidth  = (inputWidth + 2*padW - kW) / dW + 1;
+
+  Tensor input = input_;
+  if (in_sizes.size() == 3) {
+    input = input.unsqueeze(0);
+  }
+
+  // Batch size + input planes
+  in_sizes = input.sizes();
+  int64_t batchSize = in_sizes[0];
+
+  // Resize output
+  Tensor output = output_;
+  if (ndim == 4) {
+    resize_output(output, {batchSize, nOutputPlane, outputHeight, outputWidth});
+  } else {
+    resize_output(output, {nOutputPlane, outputHeight, outputWidth});
+    output = output.unsqueeze(0);
+  }
+
+  // Resize temporary columns
+  resize_output(columns, {nInputPlane * kW * kH, outputHeight * outputWidth});
+
+  // Define a buffer of ones, for bias accumulation
+  // Note: this buffer can be shared with other modules, it only ever gets increased,
+  // and always contains ones.
+  Tensor ones;
+  if (bias.defined()) {
+    ones = at::ones({outputHeight, outputWidth}, input.options());
+  }
+  const bool requires_columns = (
+      kW != 1 || kH != 1 || dW != 1 || dH != 1 || padH != 0 || padW != 0);
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(kHalf, kBFloat16, input.scalar_type(),
+                                  "slow_conv2d_cuda", [&] {
+    // For each elt in batch, do:
+    for (int elt = 0; elt < batchSize; elt ++) {
+      // Matrix mulitply per output:
+      auto input_n = input.select(0, elt);
+      auto output_n = output.select(0, elt);
+
+      // Do Bias first:
+      // M,N,K are dims of matrix A and B
+      // (see http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-gemm)
+      int64_t m_ = nOutputPlane;
+      int64_t n_ = outputHeight * outputWidth;
+      int64_t k_ = 1;
+
+      // Do GEMM (note: this is a bit confusing because gemm assumes column-major matrices)
+      if (bias.defined()) {
```

Comment:
is it just expanding bias to output size? Should we be able to do the same with `.expand()` and `output_n.copy_`, without gemm call and `ones`?

Reply:
Is it alright if I leave algorithmic changes to a follow-up PR?

Reply:
sure! Then I only have questions about 3d inputs and 2d weights.